icicle/src/poseidon/extern.cu

```diff
@@ -2,58 +2,91 @@
 
 using namespace field_config;
 
-#include "poseidon.cu"
+#include "poseidon/poseidon.cuh"
 #include "constants.cu"
 
 #include "gpu-utils/device_context.cuh"
 #include "utils/utils.h"
 
 namespace poseidon {
-  /**
-   * Extern "C" version of [poseidon_hash_cuda] function with the following
-   * value of template parameter (where the field is given by `-DFIELD` env variable during build):
-   *  - `S` is the [field](@ref scalar_t) - either a scalar field of the elliptic curve or a
-   * stand-alone "STARK field";
-   * @return `cudaSuccess` if the execution was successful and an error code otherwise.
-   */
-  extern "C" cudaError_t CONCAT_EXPAND(FIELD, poseidon_hash_cuda)(
-    scalar_t* input,
-    scalar_t* output,
-    int number_of_states,
-    int arity,
-    const PoseidonConstants<scalar_t>& constants,
-    PoseidonConfig& config)
+  template class Poseidon<scalar_t>;
```

Comment:
basically you expose all of those and then in FFI you reconstruct the hash structure and call those functions?

Reply:
yes, ffi just wraps the methods of the class